logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt

```diff
@@ -157,8 +165,30 @@ internal class ScheduleNewAssetMessageUseCaseImpl(
             // We persist the asset message right away so that it can be displayed on the conversation screen loading
             persistMessage(message).onSuccess {
                 // We schedule the asset upload and return Either.Right(Unit) so later it's transformed to Success(message.id)
-                scope.launch(dispatcher.io) {
-                    uploadAssetAndUpdateMessage(currentAssetMessageContent, message, conversationId, expectsReadConfirmation)
+                outGoingAssetUploadJob = scope.launch(dispatcher.io) {
+                    launch {
+                        messageRepository.observeMessageVisibility(message.id, conversationId).collect { visibility ->
+                            if (visibility == MessageEntity.Visibility.DELETED) {
+                                // If the message is deleted we cancel the upload
+                                outGoingAssetUploadJob?.cancel()
+                            }
+                        }
+                    }
+                    launch {
+                        uploadAssetAndUpdateMessage(currentAssetMessageContent, message, conversationId, expectsReadConfirmation)
+                    }.invokeOnCompletion { cause ->
+                        if (cause is CancellationException) {
+                            kaliumLogger.d(
+                                "Asset upload was cancelled, " +
+                                        "for message with id ${message.id} and conversationId $conversationId"
+                            )
+                        } else {
+                            kaliumLogger.d(
+                                "Asset uploaded successfully, " +
+                                        "for message with id ${message.id} and conversationId $conversationId"
+                            )
+                        }
+                    }
```

Comment:
Can we test that ?

Reply:
not sure how honestly :S any ideas ?

Reply:
Not really testable at the moment, as this `UseCase` just launches a job silently.

I believe our uploading and downloading are lacking a bit at the moment.

We could have an in-memory storage of asset upload progress, which also holds a reference to this `job` currently silently launched by this `UseCase`.

This in-memory storage could be informed that the message was deleted and then stop the upload. Or if we just want to cancel/retry asset uploads too.

Some structure like this would be needed in any case if we ever want to display upload progress properly: when displaying messages, we could pass the upload progress from this in-memory storage.

Bonus: would make this feature testable, as the job/progress is no longer hidden.

Reply:
@vitorhugods great idea I will add this as a TODO to come back to this whenever we will refine the display of upload progress